File: app/src/main/java/com/kaciula/archiman/presentation/widget/DividerItemDecoration.java
package com.kaciula.archiman.presentation.widget;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

public class DividerItemDecoration extends RecyclerView.ItemDecoration {

    private Drawable divider;
    private boolean drawLastDivider;

    public DividerItemDecoration(Context context) {
        final TypedArray a = context.obtainStyledAttributes(null, new int[]{android.R.attr
                .listDivider});
        divider = a.getDrawable(0);
        a.recycle();
        drawLastDivider = true;
    }

    public DividerItemDecoration(Context context, boolean drawLastDivider) {
        this(context);
        this.drawLastDivider = drawLastDivider;
    }

    public DividerItemDecoration(Drawable divider) {
        this.divider = divider;
        drawLastDivider = true;
    }

    @Override
    public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State
            state) {
        super.getItemOffsets(outRect, view, parent, state);
        if (divider == null) return;
        if (getOrientation(parent) == LinearLayoutManager.VERTICAL) {
            outRect.bottom = divider.getIntrinsicHeight();
        } else outRect.right = divider.getIntrinsicWidth();
    }

    @Override
    public void onDrawOver(Canvas c, RecyclerView parent, RecyclerView.State state) {
        if (divider == null) {
            super.onDrawOver(c, parent, state);
            return;
        }

        if (getOrientation(parent) == LinearLayoutManager.VERTICAL) {
            final int left = parent.getPaddingLeft();
            final int right = parent.getWidth() - parent.getPaddingRight();
            final int childCount = parent.getChildCount();

            final int count = drawLastDivider ? childCount : childCount - 1;
            for (int i = 0; i < count; i++) {
                final View child = parent.getChildAt(i);
                final RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child
                        .getLayoutParams();
                final int size = divider.getIntrinsicHeight();
                final int top = child.getBottom() + params.bottomMargin;
                final int bottom = top + size;
                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }
        } else { //horizontal
            final int top = parent.getPaddingTop();
            final int bottom = parent.getHeight() - parent.getPaddingBottom();
            final int childCount = parent.getChildCount();

            final int count = drawLastDivider ? childCount : childCount - 1;
            for (int i = 0; i < count; i++) {
                final View child = parent.getChildAt(i);
                final RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child
                        .getLayoutParams();
                final int size = divider.getIntrinsicWidth();
                final int left = child.getRight() + params.rightMargin;
                final int right = left + size;
                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }
        }
    }

    private int getOrientation(RecyclerView parent) {
        if (parent.getLayoutManager() instanceof LinearLayoutManager) {
            LinearLayoutManager layoutManager = (LinearLayoutManager) parent.getLayoutManager();
            return layoutManager.getOrientation();
        } else
            throw new IllegalStateException("DividerItemDecorationOld can only be used with a " +
                    "LinearLayoutManager.");
    }
}

File: app/src/main/kotlin/com/kaciula/archiman/presentation/widgets/DividerItemDecoration.kt
package com.kaciula.archiman.presentation.widgets

import android.content.Context
import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View

class DividerItemDecoration : RecyclerView.ItemDecoration {

    private var divider: Drawable
    private var drawLastDivider: Boolean = false

    constructor(context: Context) {
        val a = context.obtainStyledAttributes(null, intArrayOf(android.R.attr.listDivider))
        divider = a.getDrawable(0)
        a.recycle()
        drawLastDivider = true
    }

    constructor(context: Context, drawLastDivider: Boolean) : this(context) {
        this.drawLastDivider = drawLastDivider
    }

    constructor(divider: Drawable) {
        this.divider = divider
        drawLastDivider = true
    }

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView,
                                state: RecyclerView.State?) {
        super.getItemOffsets(outRect, view, parent, state)
        if (getOrientation(parent) == LinearLayoutManager.VERTICAL) {
            outRect.bottom = divider.intrinsicHeight
        } else {
            outRect.right = divider.intrinsicWidth
        }
    }

    override fun onDrawOver(canvas: Canvas, parent: RecyclerView, state: RecyclerView.State?) {
        if (getOrientation(parent) == LinearLayoutManager.VERTICAL) {
            val left = parent.paddingLeft
            val right = parent.width - parent.paddingRight
            val childCount = parent.childCount

            val count = if (drawLastDivider) childCount else childCount - 1
            for (i in 0..count - 1) {
                val child = parent.getChildAt(i)
                val params = child.layoutParams as RecyclerView.LayoutParams
                val size = divider.intrinsicHeight
                val top = child.bottom + params.bottomMargin
                val bottom = top + size
                divider.setBounds(left, top, right, bottom)
                divider.draw(canvas)
            }
        } else { //horizontal
            val top = parent.paddingTop
            val bottom = parent.height - parent.paddingBottom
            val childCount = parent.childCount

            val count = if (drawLastDivider) childCount else childCount - 1
            for (i in 0..count - 1) {
                val child = parent.getChildAt(i)
                val params = child.layoutParams as RecyclerView.LayoutParams
                val size = divider.intrinsicWidth
                val left = child.right + params.rightMargin
                val right = left + size
                divider.setBounds(left, top, right, bottom)
                divider.draw(canvas)
            }
        }
    }

    private fun getOrientation(parent: RecyclerView): Int {
        if (parent.layoutManager is LinearLayoutManager) {
            val layoutManager = parent.layoutManager as LinearLayoutManager
            return layoutManager.orientation
        } else {
            throw IllegalStateException(
                    "DividerItemDecoration can only be used with a LinearLayoutManager.")
        }
    }
}